### DISCUSSION TITLE
Installation documents for Linux users (non-developers)?

### DISCUSSION THREAD

**Discussion by kjcole** [mu-editor/mu#1426] [Q&A]
A few weeks ago, I fought with installing Mu and got it working -- or so I thought. But apparently, my notes were not as thorough as I thought.

So, once again I started with https://codewith.mu/en/howto/1.1/install_with_python and it complained about `black`:

```
ERROR: Could not find a version that matches black>=19.10b0
```

But:

```
$ pip list | grep black
black                         20.8b1
```

looks `>=19.10.b0`...

So, I came back to the git repo and tried from that. Eventually, I got it working, but I don't exactly know what I did that made it work.

The point is that someone who runs Linux, but is a new Python user should be able to follow the directions on the **Install with Python** page -- even if it's not the latest and greatest version of `Mu` and it should **just work**.

**Comment by ntoll** [comment]
Hi @kjcole - thanks for the feedback.

Out of interest, which version of Linux are you using and  (perhaps more importantly) what version of Python..?

I develop Mu on Linux (I've been a Linux user since 1997), so I'm painfully aware how difficult it can be to install packages given the complicated mix of versions, libraries, Pythons and so on. I hear ya..!

As you'll know, Linux packaging is in a state of "flux". What with flatpack, snaps, and all the rest, it's not clear which standard will become ubiquitous for the next decade (until someone re-invents packaging again, again). My general rule of thumb is to use the lightest base Linux I like (usually a Debian derivative like Ubuntu) and get the latest Python, then build up from there.

Of course, Mu is aimed at beginners for whom "open the console and create a venv" is like a slap in the face. Yet, until packaging settles down, it's hard to suggest something that *isn't* a pain in the neck in some sense.

As someone "in the field", it'd be really helpful for us to hear your thoughts on ways to move forward in this problem space... I think we can both agree that in an ideal world, a beginner Linux user would:

a) Visit codewith.mu/download
b) Download the clearly labelled Linux installer
c) Execute the installer
d) There's no an item in their launch menu labelled Mu
e) Clicking the item in the menu launches Mu

Thoughts..?

**Reply by abixadamj** [reply]
@ntoll 
My thoughts about this issue are:

We shoud focus on (in that order): 

- enable pip in pypi.python.org for those who have another distro (rpm, aur, tgz....)
- think about flatpack/snap/appimage - and choose the one, which is available for the biggest number of distributions,
- deb package for about 65% of Linux users

I am so happy, that last version is working on Linux via `run.py` https://github.com/mu-editor/mu/issues/1408#issuecomment-810467273
I can say that from april I can spent more time for helping develop Mu, so I promise to do tests first. I use Mint 20 with Python 3.8.5
```
Python 3.8.5 (default, Jan 27 2021, 15:41:15) 
[GCC 9.3.0] on linux
```
Later I can try to focus on packaging to PyPi - which could be the good point to start. 

And after that I will promise to check AppImage  (https://appimage.org/) - I think this coud be great for Linux users - just download and run ;-) Like in MacOS...

**Reply by ZanderBrown** [reply]
Some thoughts
- Snap is very ubuntu specific despite Canonical's claims, instead of “universal Linux packages” they are better described as "a deb alternative for Ubuntu by Canonical™". If Ubuntu is what you care about snaps are probably great, esp considering the rather byzantine nature of deb packaging.
- AppImage is a mess, I can see the appeal “just download and run” but that's not what they are and I'm rather uncomfortable telling users (esp those likely to be young people) to go run random binaries.
- flatpak does a good job of crossing the distro divide and flathub.org would provide good ‘visibility’ for Mu, unfortunately Ubuntu ('cos snaps) doesn't have flatpak enabled out-the-box (though the reverse is true for Fedora and I don't think Debian/Arch etc ship either)

I have previously played with flatpak'ing Mu but thus far have hit the "packaging PyQt5 is a mess" wall, that said I know it can be done since [Picard](https://www.flathub.org/apps/details/org.musicbrainz.Picard) is also PyQt5 based and I've been using it for years now

**Reply by ntoll** [reply]
OK... so "AppImage is a mess" - got any pointers as to why..? I run AppImages of various things and they make it super-duper easy... and given our users are at the "download it and just run it" stage of their careers it certainly feels like the only sort of advice we could give for any of the packaging solutions. Thoughts @ZanderBrown and anyone else with better visibility of this sort of issue than me..? (NB my level of experience is, "I've heard of them"). :slightly_smiling_face: :+1:

**Reply by kjcole** [reply]
> Hi @kjcole - thanks for the feedback.
> 
> Out of interest, which version of Linux are you using and (perhaps more importantly) what version of Python..?

* **Linux** `Ubuntu 20.04.2 LTS`
* **Kernel** `5.4.0-70-lowlatency #78-Ubuntu SMP PREEMPT  x86_64`
* **Python** `3.8.5 (default, Jan 27 2021, 15:41:15)  [GCC 9.3.0] on linux`
 
> I develop Mu on Linux (I've been a Linux user since 1997), so I'm painfully aware how difficult it can be to install packages given the complicated mix of versions, libraries, Pythons and so on. I hear ya..!
> 
> As you'll know, Linux packaging is in a state of "flux". What with flatpack, snaps, and all the rest, it's not clear which standard will become ubiquitous for the next decade (until someone re-invents packaging again, again). My general rule of thumb is to use the lightest base Linux I like (usually a Debian derivative like Ubuntu) and get the latest Python, then build up from there.

Although I recognize -- and value -- innovations and improvements (not to mention the blood, sweat and tears that go into them), and also recognize the needs of the many, one of the more important considerations when considering distributions is that I want **ONE** packaging method for the distribution. So, having `.deb` and `AppImage` and `Snap` and `flatpak` and... and... and... and `pip` and `cargo` and `npm` and.... Well, let's just say I'm old and stupid, and have been bitten in the ass by such combinations. For example, I managed to `pip upgrade` several packages one time, some of which were... oops... installed via `apt` (or maybe it was `pacman` on an Arch system). In any case it royally bollixed up my system in a way from which I never fully recovered.

Since then, I've become considerably more careful about either using `--user` with `pip` or using `pipenv` to get both a virtual environment and pip rolled nicely into one -- though that still gets me into trouble when some `apt` package, due to some configuration finds a dependency in userland before it finds it system-wide, and lo, and behold, the userland version is not what it wants. (Rare, but occasionally happens, and then I have to go add some magic environment variable. Not a huge deal, though.)

I have caved on AppImages twice, with great reluctance, but I've managed to stay away completely from snaps and flatpaks.

> Of course, Mu is aimed at beginners for whom "open the console and create a venv" is like a slap in the face. Yet, until packaging settles down, it's hard to suggest something that _isn't_ a pain in the neck in some sense.
> 
> As someone "in the field", it'd be really helpful for us to hear your thoughts on ways to move forward in this problem space... I think we can both agree that in an ideal world, a beginner Linux user would:
> 
> a) Visit codewith.mu/download
> b) Download the clearly labelled Linux installer
> c) Execute the installer
> d) There's no an item in their launch menu labelled Mu
> e) Clicking the item in the menu launches Mu

Agreed that the ideal is going to be an installer and a resulting menu entry somewhere to launch. But I was thinking of something that I hoped might be more achievable even if less ideal.

> Thoughts..?

My main thoughts are:

1. have documentation  for users that helps them determine what (if any) prerequisites / dependencies they already have installed (at what version) and recommend a corresponding version of Mu and any dependencies they might need to fulfill. 
2. change "_create a venv_" to something more like "_create a virtual environment. So-and-so has a good tutorial (insert link) on the what, why and how of virtual environments._" And / or "_Using your package manager, be sure that both the Python package manager (pip) and  the virtual environment (venv) packages for Python3 and then open a terminal and type the following..._" (with  preformatted code section that they can copy and paste from). Possibly include suggestions as to the probable name for the system-wide packages, e.g.  `python3-pip`, `python3-virtualenv`, `virtualenv`.
3. (and I don't know any good way to do it) a troubleshooting guide for beginners. Possibly with indications of what they can ignore.
4. make some dependencies optional? I mean, `black` is nice to have, but is it strictly necessary in order to get my Python code onto a micro:bit? 

I _think_ what ultimately happened was at some point with one of the methods I tried, I ended up ignoring an error message and said "Screw it. I'm forging ahead, regardless of the fact that it complained about a failed dependency.'

**Reply by ZanderBrown** [reply]
> got any pointers as to why..? I run AppImages of various things and they make it super-duper easy... and given our users are at the

So the most obvious problem is their main selling point: It's a single file you can run from xdg-downloads, so isn't installed

As a result modern platforms won't really know what to do with it, for Ubuntu GNOME/GNOME/elementary this means the app will run with no icon (or a fallback) or label in the app switcher - For that to work you need a desktop entry in `$XDG_DATA_DIRS/applications`

Another snag is that AppImages are basically a rather hole-y container so you have to bundle everything you need and in old enough versions for the places you intend to run - Generally this is achieved by building your AppImage in a VM of some LTS system. libc is a common source of fun here.

Something else to consider is the lack of any update system, users would have to check the website and download a whole new copy of Mu. flatpak and snap of course have build-in update systems, at least with flatpak these updates are even automatic and make use of deltas (which may be nice for those with patchy connections)

> "download it and just run it" stage of their careers it certainly feels like the only sort of advice we could give for any of the packaging solutions ni 

There is, esp on Linux based systems, a bit of a difference between downloading and running Mu from - say - flathub vs downloading a random file and `chmod +x`'ing it I think

**Reply by abixadamj** [reply]
Reading this, I think, that Linux is not for oridinal user. And because of so many types of packaging systems we cannot do one universal installer.

I thought about AppImage, but after rwading what @ZanderBrown wrote - I change my mind.

Now I think that the best we can do is to do:

a) DEB package to get about 60% of linux exists....
b) pip package for rest...

I can try to make it... but first vote on this.

**Reply by kjcole** [reply]
Since the Raspberry Pi is generally Linux based, with a strong emphasis on Python, I would question the contention.

But I love the idea of using Debian packages and, as an alternate pip packages. (I don't know that I get a vote, as someone who's more of a bug filer than contributor, but "+1" just in case.)

**Reply by ZanderBrown** [reply]
I mean deb and rpm packages already exist for the stable release, if we want to offer first-party packages (vs telling people to use whatever distros have made) I'd still vote for flatpak, whilst people may not have flatpak preinstalled it should be available and work everywhere

Or of course just pip everything

**Reply by 1oglop1** [reply]
Hi, I found this discussion because I tried to `pip install mu-editor` yestesterday and got bitten by missing SDL after doing the research, the first result returned `SDL2` so I happily did `apt install libsdl2-dev`.

After which I later found out is no use because `pygame < 2.0` has dependency on `SDL v1` .
My current system is `Ubuntu 21.04`.

I do not know so much about packaging but IRRC `deb` package does not contain all dependencies but instead rely on downloading dependency from package repository.

Therefore I'm in favour of `snap` or `flatpack` ideally both since both of them support incremental updates and in case you get an inexperienced linux user, there is very high chance they will be using ubuntu so installing MU from the Software center is great option.

**Reply by carlosperate** [reply]
The latest releases of Mu have been betas, so they are marked as "pre-release" in PyPI. You can use the `--pre` flag to pip install the pre-release version, which uses pygame v2, so you can specify a version like `pip install mu-editor==1.1.0b4`.

That should get you one step closer.

Was there any Mu documentation that indicated to run `pip install mu-editor` instead of `pip install mu-editor==1.1.0b4`, if that's the case it would be great to know so that we can ensure the docs are up-to-date.

**Reply by 1oglop1** [reply]
> The latest releases of Mu have been betas, so they are marked as "pre-release" in PyPI. You can use the `--pre` flag to pip install the pre-release version, which uses pygame v2, so you can specify a version like `pip install mu-editor==1.1.0b4`.
> 
> That should get you one step closer.
> 
> Was there any Mu documentation that indicated to run `pip install mu-editor` instead of `pip install mu-editor==1.1.0b4`, if that's the case it would be great to know so that we can ensure the docs are up-to-date.

It was my bad, when I saw `==XXb` I automatically ignored it and went for the `pip install mu-editor` version - skipping a lot of text around as I was a little bit in a time pressure to prepare.

**Comment by abixadamj** [comment]
Just for the record... I have installed fresh clean Ubuntu 20.04 LTS yesterday and now I am trying to run mu.
What need to be installed first:
```
sudo apt install python3-pyqt5
sudo apt install python3-appdirs
sudo apt install python3-pyflakes
sudo apt install python3-pycodestyle
sudo apt install python3-qtconsole
sudo apt install python3-pyqt5.qsci
sudo apt install python3-serial
sudo apt install python3-pyqt5.qtserialport
sudo apt install python3-semver
sudo apt install python3-virtualenv
sudo -H pip3  install adafruit_board_toolkit
```
After that `python3 run.py` works perfectly.
![Zrzut ekranu z 2021-06-21 14-54-10](https://user-images.githubusercontent.com/5088643/122765208-91f0a680-d2a0-11eb-979c-722fcd86513e.png)

**Reply by carlosperate** [reply]
Technically you shouldn't have to `apt install` any of those packages in Ubuntu 20.04 (as long as it isn't running on Arm).

All the wheels should be available in PyPI for this platform, so creating a virtual environment (and maybe updating pip) should be enough for `pip install mu-editor==1.1.0b4` to work, no?

**Reply by abixadamj** [reply]
With fresh Ubuntu install, I don't have `vurtualenv`, so I must install it via `apt`. I will try it later and write more info.

**Comment by tmontes** [comment]
All,

First attempt at AppImage Linux packaging for the current release of Mu just automatically built available [at the bottom of this page](https://github.com/mu-editor/mu/actions/runs/2024827478).

Just tested it on Debian 10:
* Download.
* Make executable.
* Run it!

Wondering if the Linux users here would want to take it for a spin and see if it works for them. :-)
